jwt/jwt.go

```diff
@@ -0,0 +1,156 @@
+	// it's expired. Useful for defending against server clock skews.
+	Leeway time.Duration
+	// IsSatellite signifies that the JWT is verified on a satellite domain.
+	IsSatellite bool
+	// ProxyURL is the URL of the server that proxies the Clerk Frontend API.
+	ProxyURL *string
+	// List of values that should match the azp claim.
+	// Use SetAuthorizedParties to set the value.
+	authorizedParties map[string]struct{}
+}
+
+// SetAuthorizedParties accepts a list of authorized parties to be
+// set on the params.
+func (params *VerifyParams) SetAuthorizedParties(parties ...string) {
+	azp := make(map[string]struct{})
+	for _, p := range parties {
+		azp[p] = struct{}{}
+	}
+	params.authorizedParties = azp
+}
+
+// Verify verifies a Clerk session JWT and returns the parsed
+// clerk.SessionClaims.
+func Verify(ctx context.Context, params *VerifyParams) (*clerk.SessionClaims, error) {
+	parsedToken, err := jwt.ParseSigned(params.Token)
+	if err != nil {
+		return nil, err
+	}
+	if len(parsedToken.Headers) == 0 {
+		return nil, fmt.Errorf("missing jwt headers")
+	}
+
+	kid := parsedToken.Headers[0].KeyID
+	if kid == "" {
+		return nil, fmt.Errorf("missing jwt kid header claim")
+	}
+
+	jwk := params.JWK
+	if jwk == nil {
+		jwk, err = getJWK(ctx, kid)
+		if err != nil {
+			return nil, fmt.Errorf("get jwk: %w", err)
+		}
+	}
+
+	if parsedToken.Headers[0].Algorithm != jwk.Algorithm {
+		return nil, fmt.Errorf("invalid signing algorithm %s", jwk.Algorithm)
+	}
+
+	claims := &clerk.SessionClaims{}
+	allClaims := []any{claims}
+	if params.CustomClaims != nil {
+		allClaims = append(allClaims, params.CustomClaims)
+	}
+	err = parsedToken.Claims(jwk.Key, allClaims...)
+	if err != nil {
+		return nil, err
+	}
+
+	err = claims.Claims.ValidateWithLeeway(jwt.Expected{Time: time.Now().UTC()}, params.Leeway)
+	if err != nil {
+		return nil, err
+	}
+
+	iss := claims.Issuer
+	if params.ProxyURL != nil && *params.ProxyURL != "" {
+		iss = *params.ProxyURL
+	}
+	// Non-satellite domains must validate the issuer.
+	if !params.IsSatellite && !isValidIssuer(iss) {
+		return nil, fmt.Errorf("invalid issuer %s", iss)
+	}
+
+	if claims.AuthorizedParty != "" && len(params.authorizedParties) > 0 {
+		if _, ok := params.authorizedParties[claims.AuthorizedParty]; !ok {
+			return nil, fmt.Errorf("invalid authorized party %s", claims.AuthorizedParty)
+		}
+	}
+
+	return claims, nil
+}
+
+func getJWK(ctx context.Context, kid string) (*clerk.JSONWebKey, error) {
+	// TODO Avoid multiple requests by caching results for the same
+	// instance.
+	jwks, err := jwks.Get(ctx, &jwks.GetParams{})
+	if err != nil {
+		return nil, err
+	}
+	for _, k := range jwks.Keys {
+		if k.KeyID == kid {
+			return &k, nil
+		}
+	}
+	return nil, fmt.Errorf("no jwk key found for kid %s", kid)
+}
+
+func isValidIssuer(iss string) bool {
+	return strings.HasPrefix(iss, "https://clerk.") ||
+		strings.Contains(iss, ".clerk.accounts")
+}
+
+type DecodeParams struct {
+	Token string
```

Comment:
❓ What's the reason for wrapping `Token` in a struct? Is it for providing us a way to add more options without breaking compatibility (i.e. keeping the function signature intact)?

If so, I'm having a hard time imagining something that `Decode` would need to be extended with, that a functional option wouldn't be a fit for. In other words, perhaps it's perfectly fine to go with `Decode(ctx, string)` and in the future extend with functional options if needed? The reason I'm suggesting this is because the latter seems a simpler/easier to use API to me.

Reply:
Yes, we're trying to use custom types for method arguments so that we can easily extend them in the future without introducing breaking changes in the API.